src/mono/mono/mini/aot-runtime.c

```diff
@@ -2439,13 +2439,15 @@ static void
 load_container_amodule (MonoAssemblyLoadContext *alc)
 {
 	// If container_amodule loaded, don't lock the runtime
-	if (container_amodule)
-		return;
+    if (!container_assm_name || container_amodule)
+        return;
 
 	mono_loader_lock ();
 	// There might be several threads that passed the first check
 	// Adding another check to ensure single load of a container assembly due to race condition 
 	if (!container_amodule) {
+        char *local_ref = container_assm_name;
```

Comment:
This shouldn't be needed any more.
The locking should be:
if (container_amodule)
   return;
mono_loader_lock ();
if (!container_amodule) {
   <rest>
  mono_memory_barrier ();
  container_amodule = assm;
}
mono_loader_unlock ();

Reply:
It looks like threads are not in sync without `container_assm_name` check.

```c
if (!container_assm_name || container_amodule)
		return;

char *local_ref = container_assm_name;
container_assm_name = NULL;
```

The library tests fail at `g_assert (assm->image->aot_module);` as it is NULL.

Reply:
It seems that the following is used to prevent recursive calls within the same thread.
```c
char *local_ref = container_assm_name;
container_assm_name = NULL;
```

When `load_container_amodule` is invoked initially, `load_aot_module` is invoked with the fake container assembly and image. `load_aot_module` invokes `load_image` that indirectly invokes `mono_aot_get_method` for `OnLoadAssembly` method. Then, `load_container_amodule` is invoked again within the same thread, and assumption is that it can enter the lock.

As `OnLoadAssembly` method is not emitted into the container AOT module, it is not needed to get the method.

Reply:
Yes, its still needed, load_aot_module () is not thread safe since it assumes its being called from the assembly loader code which does all kinds of locking.